extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java

```diff
@@ -0,0 +1,76 @@
+package io.quarkus.infinispan.client.runtime.cache;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.CacheException;
+
+import io.smallrye.mutiny.Uni;
+
+public class InfinispanGetWrapper {
+    final RemoteCache cache;
+    private final Map<Object, ReentrantLock> synchronousGetLocks = new ConcurrentHashMap<>();
+
+    public InfinispanGetWrapper(RemoteCache cache) {
+        this.cache = cache;
+    }
+
+    public <K, V> Uni<V> get(K key, Function<K, V> valueLoader) {
+        return Uni.createFrom().completionStage(
+                /*
+                 * Even if CompletionStage is eager, the Supplier used below guarantees that the cache value computation will be
+                 * delayed until subscription time. In other words, the cache value computation is done lazily.
+                 */
+                new Supplier<CompletionStage<V>>() {
+                    @Override
+                    public CompletionStage<V> get() {
+                        CompletionStage<Object> infinispanValue = getFromInfinispan(key, valueLoader);
+                        return cast(infinispanValue);
+                    }
+                });
+    }
+
+    private <K, V> CompletableFuture<Object> getFromInfinispan(K key, Function<K, V> valueLoader) {
+        ReentrantLock lock;
+        Object value = cache.get(key);
+        if (value == null) {
+            lock = synchronousGetLocks.computeIfAbsent(key, k -> new ReentrantLock());
+            lock.lock();
```

Comment:
Why is the lock logic needed here? A new `InfinispanGetWrapper` instance is created on each `CacheResultInterceptor#intercept` invocation so I can't see when this lock will actually prevent a cache value from being stored in the remote cache using the value loader.

The lock logic would make more sense to me if the locks were stored in a "persistent" (from a memory perspective) wrapper instance around `RemoteCache`, but maybe I missed something.

Reply:
You are right on this, since the object is created each time.
This is me extracting the code and then not realizing on this part haha thanks for that
we need the lock from the client perspective as we do in the infinispan spring-boot integration.
let me fix this and ping you back 
thanks!